File: openstudio-necb/lib/openstudio_necb/report/checklist.rb
require 'set'

module OpenStudioNECB
  module Report
    # Derives the AHJ-style checklist from the audit log. Rows come from
    # :compliance decisions (the article-cited verdicts), and every warning is
    # elevated so the checklist can never look cleaner than the run actually
    # was. Coverage entries are reconciled: a host_scope delegation with no
    # implementing coverage entry in the same audit surfaces as a warning row;
    # everything else stays in the coverage appendix.
    module Checklist
      Row = Struct.new(:glyph, :article, :statement, :measured, :audit_index, :building, keyword_init: true)

      # The audit convention SHOUTS violations ('EXCEEDS', 'does NOT meet',
      # 'BELOW the') while pass texts stay lowercase ('does not exceed',
      # 'within 100 h') — so the fail check is deliberately case-SENSITIVE.
      FAIL_WORDS = /\b(EXCEEDS?|NOT|BELOW|FAILS?)\b|non-compliant/.freeze
      PASS_WORDS = /\b(meets|complies|compliant|does not exceed|within|satisfied)\b/i.freeze

      # Coverage statuses that count as an article being actually handled in
      # this run (so a sibling host_scope delegation is reconciled, not warned).
      COVERING_STATUSES = %w[implemented partial satisfied_by_clone].freeze

      module_function

      # @param audit_entries [Array<Hash>] AuditLog#entries (symbol keys)
      # @return [Array<Row>] article-sorted checklist rows
      def rows(audit_entries)
        covered = covered_articles(audit_entries)
        out = []
        audit_entries.each_with_index do |entry, index|
          step = entry[:step].to_s
          inputs = entry[:inputs]
          status = inputs.is_a?(Hash) ? inputs[:status].to_s : ''
          if %i[warn warning].include?(entry[:level])
            out << Row.new(glyph: :warning, article: entry[:article].to_s, building: entry[:building],
                           statement: statement_for(entry), measured: measured_for(entry), audit_index: index)
          elsif step == 'compliance' && entry[:level] == :decision
            out << Row.new(glyph: verdict_glyph(entry[:action].to_s), article: entry[:article].to_s,
                           building: entry[:building],
                           statement: statement_for(entry), measured: measured_for(entry), audit_index: index)
          elsif step == 'coverage' && entry[:level] == :info && status == 'host_scope' &&
                !covered?(entry[:article].to_s, covered)
            out << Row.new(glyph: :warning, article: entry[:article].to_s, building: entry[:building],
                           statement: "Delegated but NOT covered in this run: #{entry[:action]}",
                           measured: nil, audit_index: index)
          end
        end
        out.sort_by { |r| [article_sort_key(r.article), r.audit_index] }
      end

      # Articles an implementing coverage entry claims in THIS audit. Shared by
      # the checklist reconciliation and the coverage appendix so both agree.
      def covered_articles(audit_entries)
        covered = Set.new
        audit_entries.each do |entry|
          next unless entry[:step].to_s == 'coverage'

          inputs = entry[:inputs]
          next unless inputs.is_a?(Hash) && COVERING_STATUSES.include?(inputs[:status].to_s)

          article = entry[:article].to_s
          covered << article unless article.empty?
        end
        covered
      end

      # Bidirectional prefix match (subsumes exact equality): '8.4.4.20.' covers
      # '8.4.4.20.(1)' and vice versa. NECB article strings end in '.', so the
      # trailing dot guards against '8.4.4.2.' matching '8.4.4.20.'.
      def covered?(article, covered_set)
        a = article.to_s
        return false if a.empty?

        covered_set.any? { |b| a.start_with?(b) || b.start_with?(a) }
      end

      def verdict_glyph(action)
        return :fail if action =~ FAIL_WORDS
        return :pass if action =~ PASS_WORDS

        :na
      end

      def statement_for(entry)
        text = entry[:action].to_s
        target = entry[:target]
        target && !text.include?(target.to_s) ? "#{target}: #{text}" : text
      end

      # Compact "measured" cell from the decision's inputs hash.
      def measured_for(entry)
        inputs = entry[:inputs]
        return nil unless inputs.is_a?(Hash) && !inputs.empty?

        inputs.first(4).map do |k, v|
          value = v.is_a?(Float) ? v.round(2) : v
          "#{k}: #{value}"
        end.join(', ')
      end

      # Sort "8.4.1.2.(2)" numerically per level; unknown articles sink last.
      def article_sort_key(article)
        numbers = article.to_s.scan(/\d+/).map(&:to_i)
        numbers.empty? ? [99] : numbers
      end
    end
  end
end

File: python/btap/necb/report/checklist.py
"""Derives the AHJ-style checklist from the audit log (port of
report/checklist.rb). Rows come from 'compliance' decisions (the
article-cited verdicts), and every warning is elevated so the checklist can
never look cleaner than the run actually was. Coverage entries are
reconciled: a host_scope delegation with no implementing coverage entry in
the same audit surfaces as a warning row; everything else stays in the
coverage appendix.

Port note (D-79): the Python AuditLog records str-keyed entries with string
levels ('decision' | 'info' | 'warning'), so the membership tests here read
those spellings — Ruby's ``%i[warn warning]`` both landed as :warning in the
persisted audit anyway."""

from __future__ import annotations

import re
from dataclasses import dataclass

from btap._compat import ruby_round, ruby_str

#: The audit convention SHOUTS violations ('EXCEEDS', 'does NOT meet',
#: 'BELOW the') while pass texts stay lowercase ('does not exceed',
#: 'within 100 h') — so the fail check is deliberately case-SENSITIVE.
FAIL_WORDS = re.compile(r"\b(EXCEEDS?|NOT|BELOW|FAILS?)\b|non-compliant")
PASS_WORDS = re.compile(
    r"\b(meets|complies|compliant|does not exceed|within|satisfied)\b",
    re.IGNORECASE)

#: Coverage statuses that count as an article being actually handled in this
#: run (so a sibling host_scope delegation is reconciled, not warned).
COVERING_STATUSES = ("implemented", "partial", "satisfied_by_clone")


@dataclass
class Row:
    glyph: str
    article: str
    statement: str
    measured: str | None
    audit_index: int
    building: str | None = None


def rows(audit_entries):
    """:param audit_entries: AuditLog.entries (str-keyed dicts)
    :return: article-sorted checklist Rows"""
    covered = covered_articles(audit_entries)
    out = []
    for index, entry in enumerate(audit_entries):
        step = str(entry.get("step") or "")
        inputs = entry.get("inputs")
        status = str(inputs.get("status") or "") if isinstance(inputs, dict) else ""
        level = entry.get("level")
        if level == "warning":
            out.append(Row(glyph="warning", article=_article(entry),
                           building=entry.get("building"),
                           statement=statement_for(entry),
                           measured=measured_for(entry), audit_index=index))
        elif step == "compliance" and level == "decision":
            out.append(Row(glyph=verdict_glyph(str(entry.get("action") or "")),
                           article=_article(entry),
                           building=entry.get("building"),
                           statement=statement_for(entry),
                           measured=measured_for(entry), audit_index=index))
        elif (step == "coverage" and level == "info" and status == "host_scope"
              and not covered_by(_article(entry), covered)):
            out.append(Row(glyph="warning", article=_article(entry),
                           building=entry.get("building"),
                           statement="Delegated but NOT covered in this run: "
                                     f"{entry.get('action')}",
                           measured=None, audit_index=index))
    return sorted(out, key=lambda r: (article_sort_key(r.article), r.audit_index))


def covered_articles(audit_entries):
    """Articles an implementing coverage entry claims in THIS audit. Shared
    by the checklist reconciliation and the coverage appendix so both
    agree."""
    covered = set()
    for entry in audit_entries:
        if str(entry.get("step") or "") != "coverage":
            continue

        inputs = entry.get("inputs")
        if not (isinstance(inputs, dict)
                and str(inputs.get("status") or "") in COVERING_STATUSES):
            continue

        article = _article(entry)
        if article:
            covered.add(article)
    return covered


def covered_by(article, covered_set):
    """Bidirectional prefix match (subsumes exact equality): '8.4.4.20.'
    covers '8.4.4.20.(1)' and vice versa. NECB article strings end in '.',
    so the trailing dot guards against '8.4.4.2.' matching '8.4.4.20.'."""
    a = str(article or "")
    if not a:
        return False

    return any(a.startswith(b) or b.startswith(a) for b in covered_set)


def verdict_glyph(action):
    if FAIL_WORDS.search(action):
        return "fail"
    if PASS_WORDS.search(action):
        return "pass"

    return "na"


def statement_for(entry):
    text = str(entry.get("action") or "")
    target = entry.get("target")
    if target is not None and str(target) not in text:
        return f"{target}: {text}"
    return text


def measured_for(entry):
    """Compact "measured" cell from the decision's inputs dict."""
    inputs = entry.get("inputs")
    if not (isinstance(inputs, dict) and inputs):
        return None

    parts = []
    for k, v in list(inputs.items())[:4]:
        value = ruby_round(v, 2) if isinstance(v, float) else v
        parts.append(f"{k}: {ruby_str(value)}")
    return ", ".join(parts)


def article_sort_key(article):
    """Sort "8.4.1.2.(2)" numerically per level; unknown articles sink
    last."""
    numbers = [int(n) for n in re.findall(r"\d+", str(article or ""))]
    return numbers if numbers else [99]


def _article(entry):
    return str(entry.get("article") or "")
